SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift

```diff
@@ -539,6 +560,63 @@ extension Function {
     }
     return nil
   }
+
+  var mayBindDynamicSelf: Bool {
+    self.bridged.mayBindDynamicSelf()
+  }
+
+  enum NonFailureExitBBsSearchResult {
+    case allBlocksNotUnderstood
+    case allBlocksUnderstood([BasicBlock])
+  }
+
+  func findAllNonFailureExitBBs() -> NonFailureExitBBsSearchResult {
+    var nonFailureExitBBs: [BasicBlock] = []
+
+    for block in self.blocks {
+      let termInst = block.terminator
+
+      // If we know that this terminator is not an exit terminator, continue.
+      if termInst.isSafeNonExitTerminator {
+        continue
+      }
+
+      // A return inst is always a non-failure exit bb.
+      if termInst.isFunctionExiting {
+        nonFailureExitBBs.append(block)
+        continue
+      }
+
+      // If we don't have an unreachable inst at this point, this is a terminator
+      // we don't understand. Be conservative and return false.
+      if !(termInst is UnreachableInst) {
+        return .allBlocksNotUnderstood
```

Comment:
There are not so many terminator instructions. Let's just handle them all then we don't need `allBlocksNotUnderstood`.

Maybe it's even better to replace this confusing if-cascade with a switch over all terminator instructions.